component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/MoesifClient.java

```diff
@@ -0,0 +1,206 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.am.analytics.publisher.client;
+
+import com.google.gson.Gson;
+import com.moesif.api.MoesifAPIClient;
+import com.moesif.api.controllers.APIController;
+import com.moesif.api.http.client.APICallBack;
+import com.moesif.api.http.client.HttpContext;
+import com.moesif.api.http.response.HttpResponse;
+import com.moesif.api.models.EventBuilder;
+import com.moesif.api.models.EventModel;
+import com.moesif.api.models.EventRequestBuilder;
+import com.moesif.api.models.EventRequestModel;
+import com.moesif.api.models.EventResponseBuilder;
+import com.moesif.api.models.EventResponseModel;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.am.analytics.publisher.exception.MetricReportingException;
+import org.wso2.am.analytics.publisher.reporter.MetricEventBuilder;
+import org.wso2.am.analytics.publisher.reporter.moesif.MissedEventHandler;
+import org.wso2.am.analytics.publisher.reporter.moesif.util.MoesifMicroserviceConstants;
+import org.wso2.am.analytics.publisher.retriever.MoesifKeyRetriever;
+import org.wso2.am.analytics.publisher.util.Constants;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Timer;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Moesif Client is responsible for sending events to
+ * Moesif Analytics Dashboard.
+ */
+public class MoesifClient {
+    private final Logger log = LoggerFactory.getLogger(MoesifClient.class);
+    private final MoesifKeyRetriever keyRetriever;
+    private final MissedEventHandler missedEventHandler;
+
+    public MoesifClient(MoesifKeyRetriever keyRetriever) {
+        this.keyRetriever = keyRetriever;
+        this.missedEventHandler = new MissedEventHandler(keyRetriever);
+        // execute MissedEventHandler periodically.
+        Timer timer = new Timer();
+        timer.schedule(missedEventHandler, 0, MoesifMicroserviceConstants.PERIODIC_CALL_DELAY);
+    }
+
+    private void doRetry(String orgId, MetricEventBuilder builder) {
+        Integer currentAttempt = MoesifClientContextHolder.PUBLISH_ATTEMPTS.get();
+
+        if (currentAttempt > 0) {
+            if (currentAttempt == MoesifMicroserviceConstants.NUM_RETRY_ATTEMPTS_PUBLISH) {
+                // on failure remove the respective moesif key from the map.
+                // this will result in a call to the microservice to retrieve the key.
+                // This is enough to happen only at the first attempt.
+                keyRetriever.removeMoesifKeyFromMap(orgId);
```

Comment:
AFAIU, this code block is only executed in case of an error with status code 5xx. In such scenarios, they are server side errors and removing moesif key won't help anyway.